src/util.cc

```diff
@@ -2414,10 +2441,17 @@ string json_escape_utf8(const string& val)
 	auto val_size = val.length();
 
 	// Reserve at least the size of the existing string to avoid resizing the string in the best-case
-	// scenario where we don't have any multi-byte characters.
-	string result;
-	result.reserve(val_size);
-
+	// scenario where we don't have any multi-byte characters. We keep two versions of this string:
+	// one that has a valid utf8 string and one that has a fully-escaped version. The utf8 string gets
+	// returned if all of the characters were valid utf8 sequences, but it will fall back to the
+	// escaped version otherwise. This uses slightly more memory but it avoids looping through all
+	// of the characters a second time in the case of a bad utf8 sequence.
+	string utf_result;
+	utf_result.reserve(val_size);
+	string escaped_result;
+	escaped_result.reserve(val_size);
+
+	bool found_bad = false;
 	size_t idx;
 	for ( idx = 0; idx < val_size; )
```

Comment:
Suggest to not use a for-loop here, it's easy to miss that there's no auto-increment:
```suggestion
	size_t idx = 0;
	while ( idx < val_size )
```